src/widget/wtrackmenu.cpp

```diff
@@ -1528,6 +1549,127 @@ void WTrackMenu::slotClearAllMetadata() {
             &trackOperator);
 }
 
+namespace {
+
+class RemoveTrackFilesFromDiskTrackPointerOperation : public mixxx::TrackPointerOperation {
+  public:
+    explicit RemoveTrackFilesFromDiskTrackPointerOperation(
+            TrackCollectionManager* trackCollectionManager)
+            : m_pTrackCollectionManager(trackCollectionManager) {
+    }
+
+  private:
+    void doApply(
+            const TrackPointer& pTrack) const override {
+        QString location = pTrack->getLocation();
+        qDebug() << "_";
+        qDebug() << "   RemoveTrackFilesFromDiskTrackPointerOperation";
+        qDebug() << "       " << location;
+        qDebug() << "_";
+        if (location.isNull()) {
+            return;
+        }
+        QFile file(location);
+        auto trackRef = TrackRef::fromFileInfo(
+                pTrack->getFileInfo(),
+                pTrack->getId());
+        if (!trackRef.isValid()) {
+            return;
+        }
+        QList<TrackRef> trackRefs;
+        // QList<TrackId> trackIds;
+        // TrackId id = pTrack->getId();
+        if (!file.exists()) {
+            // Missing file, purge.
+            trackRefs.append(trackRef);
+            m_pTrackCollectionManager->purgeTracks(trackRefs);
+            // m_pTrackCollectionManager->hideTracks(trackIds);
+            // trackIds.append(id);
+            return;
+        }
+        // TODO Send dialog return to purgeTracks
+        // TODO Also show reason of failure?
+        if (!file.remove()) {
+            return;
+        } else {
+            // successfully deleted, now purge.
+            trackRefs.append(trackRef);
+            m_pTrackCollectionManager->purgeTracks(trackRefs);
+            // m_pTrackCollectionManager->hideTracks(trackIds);
+            // trackIds.append(id);
+            return;
+        }
+    }
+
+    TrackCollectionManager* m_pTrackCollectionManager;
+};
+
+} // anonymous namespace
+
+void WTrackMenu::slotRemoveFromDisk() {
+    // Remove duplicates to prevent delete errors later.
+    // Update: only applies to file list in dialog,
+    // deletion is handled in independet batch operation
+    QList<TrackRef> trackRefsToDelete;
```

Comment:
Too much duplication. Only collect the locations in a QSet. Or even more efficient use a pre-allocated vector, sort, and then dedup.

 But collecting and displaying all TrackRefs upfront is still an ugly hack.

Reply:
The Delete action should be handled carefully and to avoid irrevocable mistakes we need to show the queued files.
I'll find a more elegant way to do this.